Display inventory items cost backend

ScarpMetal opened the pull request:
Do not review until PR https://github.com/polymerteam/wafflecone/pull/210 is merged into the cogs branch. 

The changes in this PR are used in order to display aggregate cost totals on the inventory screen. It would be very helpful if someone could closely review my queries and see if there is a more efficient way of accessing this data. There are some weird nuances to aggregating cost when there are inventory adjustments, so the solution was no obvious to me.

ScarpMetal committed:
Backend for aggregate inventory filtering

Most notable change is that the serializer now returns process_type and product_types which are a dictionary and an array respectively. This further reflects how ActivityList is structured and will make implementing new features in each dashboard easier in the future.

ScarpMetal committed:
Aggregate bar backend: aggregate by process_type

ScarpMetal committed:
Removed stray comments

mayanb commented on ics/v11/queries/inventory.py:
```diff
@@ -39,3 +39,44 @@ def old_inventory_used_amount(items_query):
+			cost=F('creating_task__cost')
+		).order_by('-created_at')
+
+	for item in items:
```
https://stackoverflow.com/questions/43517901/cumulative-running-sum-with-django-orm-and-postgresql
^i think we're trying to do something like this where we have a cumulative sum on the item's amount and we are partitioning on that amount - there seems to be a way here to do that in a single django query; not sure if you've already seen/tried this.
also im assuming this is an initial version and we will be updating the value of an inventory unit later to include the cost of creating minus the cost used?

mayanb commented on ics/v11/serializers.py:
```diff
@@ -906,66 +906,99 @@ class Meta:
+			'category': item['creating_task__process_type__category'],
+		}
+	
+	def get_product_types(self, item_summary):
```
if we're using this same exact function in multiple places we should probably pull it out

ScarpMetal replied:
Do you think it would be worth it to have both InventoryList2 and InventoryList2Aggregate (views.py) both use InventoryList2Serializer (rather than each having their own)?

ScarpMetal committed:
Updated get_adjusted_item_cost to use a single Django query and consolidated InventoryList2Serializer and InventoryList2Serializer into a single serializer

ScarpMetal committed:
removed code that returned incorrect inventory adjustments

userprofile__team=self.team was causing there to be an incorrect queryset of inventory adjustments. This is an unnecessary filtering because teams do not share product or process types.

mayanb commented on ics/v11/queries/inventory.py:
```diff
@@ -39,3 +39,55 @@ def old_inventory_used_amount(items_query):
+				order_by='ics_item.created_at DESC'
+			),
+			cumulative_cost=Func(
+				Sum('creating_task__cost'), 
```
we should use the cost field we defined earlier here so that later we can edit the cost field to be creating cost - remaining cost and we only have to change it in that one place

mayanb replied:
we also need to handle the case here where the cost is null- right now it's going to break.

mayanb reviewed (changes requested):
this currently breaks if any cost is null

ScarpMetal committed:
Account for costs even when they are NoneType

Also refactored and commented some code

ScarpMetal committed:
Fixed calculation issue that would give us incorrect costs

ScarpMetal committed:
Performance Improvements

There is a 12.3% decrease in runtime using this modified version of the get_adjusted_item_cost function (tested 20000+ times)

mayanb commented on ics/v11/queries/inventory.py:
```diff
@@ -39,3 +39,83 @@ def old_inventory_used_amount(items_query):
+			cost_per_unit = 0
+		cost_offset = item.amount_diff*cost_per_unit
+
+	#total_cumulative_cost = item.cumulative_cost + item.cumulative_cost_offset
```
can we remove this line?
